### ISSUE TITLE
Issue connecting by DNS name with ".connect"

### ISSUE THREAD

**Issue by fangme** [paramiko/paramiko#1712]
I have a small program I'm trying to use and connect to a linux/unix box with SSHClient.   In this specific case, the remote system is a Raspberry Pi running raspbian (which I'm pretty sure has NO bearing on the issue).

My local network has a caching DNS that I use for internal addresses - I do not modify /etc/hosts, etc.  In this case, ARP is used for the system to announce itself, and the caching DNS on my router will cache the name and IP as <systemname>.local

I can always drop to a shell, and use the installed ssh client to connect to this system using ssh pi@<systemname>.local without fail.

However, if I try to connect using:
    client.connect('<systemname>.local', username=user, password=pass, port=22)
will ALWAYS timeout.

If I use the IP address:
    client.connect('10.0.0.50', username=user, password=pass, port=22)
it will IMMEDIATELY connect without issue.

If I modify this slightly like:
    dns = socket.gethostbyname("<systemname>.local")
    client.connect(dns.strip', username=user, password=pass, port=22)
It will again immediately connect without issue

So this isn't a case of python not being able to resolve system names within it's libraries - this appears to be something within paramiko itself.

Here's the full sample code:
import paramiko
import socket
import getpass
import sys

dns = socket.gethostbyname("somesystem.local")

user="<username>"
pass="<password>"

client = paramiko.SSHClient()
client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

try:
#    client.connect('10.0.0.53', username=user, password=pass, port=22)
#    client.connect(socket.gethostbyname("somesystem.local"), username=user, password=pass, port=22)
    client.connect(dns.strip(), username=user', password=pass, port=22)
    print("Success on connection")
except paramiko.SSHException:
    print("Connection Failed")
    quit()

stdin, stdout, stderr = client.exec_command("ls -la; df -h")
for line in stdout.readlines():
    print (line.strip())

client.close()

del client, stdout, stderr
sys.exit()

sys.exit()

**Comment by MrMino** [comment]
Please format the code so that the issue is readable.
Could you provide the whole traceback for the timeout issues?

Paramiko does not use `gethostbyname()`, it loops through possible socket options for a given address, and uses the first that works. The data for that is gathered using `getaddrinfo()` (nearly the same as `socket.create_connection`). Please check whether that can resolve your domain.

**Comment by fangme** [comment]
I'm looking into Python - basically trying some various things BEFORE getting into it too deeply.  I need to make sure it can handle doing certain things before even considering converting what we have into something more consistent (like python).  Right now, stuff is scattered over 3 languages, multiple versions, etc - I'm wanting to clean that up and consolidate it.  SO - I'm not familiar yet with various things in python (specifically - python3).

1.  I've tried to clean up the source example I posted.  Hopefully that will do

2.  I'm not sure how to create a whole traceback.

3.  To answer your third question, I created:
````python
import paramiko
import socket
import getpass

addrinfo = socket.getaddrinfo("pimachine.local", 22)
print (addrinfo)

````
which returned:
[(<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('fe80::15a5:e39d:2ab0:26f', 22, 0, 4)), (<AddressFamily.AF_INET6: 30>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('fe80::15a5:e39d:2ab0:26f', 22, 0, 4)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_DGRAM: 2>, 17, '', ('10.0.0.53', 22)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('10.0.0.53', 22))]

so it seems to be getting the address (10.0.0.53), but doesn't use it with the above listed code

I am using python3 on a mac, version 3.8.1 (v3.8.1:1b293b6006)

**Comment by MrMino** [comment]
Have you tried to connect to the first one in the list?

**Comment by fangme** [comment]
No, I can do so sometime in the next day, but don't understand why this won't connect:

````python
client.connect("machinename.local", username=user, password=pass, port=22)

````

socket.socket(*addrinfo[1][:3]).connect(addrinfo[1][4])

I've tried the above - syntax errors, and right now don't have time to research.  As I said, I'm researching, and if simple things that should work don't - I don't have a lot of time to spare to figure it out.  If you could put that in the context of my original code I'd be happy to try it out.  There appears to be no left paren to properly pass the parms to the "connect".

Don't know if anyone has ever put this in before - but there are several of these I've found that are having the same original issue I've had going back to Python 2.

**Comment by MrMino** [comment]
What does the `SyntaxError` say? Which python version you're using? What's the full console output from the interpreter when you get the error? 

Well... if you're not willing to show the effort (or at least give proper logs) please don't expect others to do it for you. With all due respect, given your responses, it kind of feels like you're trying to put me in the category of "environment bugs hunting service" kind of people. I'm not - at least not for free.

I'm just a bystander who is using the library in my day to day work, and that's the first time I've seen issue like this. That's why I'm trying to help - out of simple curiosity.

There's no magic under the hood here. It's a simple `getaddrinfo()`->`s = socket.socket(...)` -> `s.connect(...)`. If that doesn't work for you, it means that you have your DNS misconfigured in your OS (most likely), or your DNS is giving out IPv6 addresses when there's no IPv6 route to the device (likely). It's as simple as that.

On the other hand, if it *does* work for you, then we might start to look for the issue in the configuration of your SSH server (somewhat likely) or Paramiko (unlikely).

**Comment by ploxiln** [comment]
I suggest trying to use https://github.com/paramiko/paramiko/pull/1405 and seeing if that works.

These "machine.local" dns names are _often_ not regular dns, they are mdns aka multicast-dns aka bonjour (on macOS) aka avahi (on linux) addresses. They _should_ work transparently ... but the old paramiko code which iterates through addresses and families might not do as good a job as python's `socket.create_connection()`.

Just an idea.

**Comment by fangme** [comment]
Thanks to both of you.  Right now, due to this little thing called Covid-19, I'm swamped, and have been for the last nearly 4 months - like others.  Having to hop between 3 different versions of Visual Studio, 3 different .NET languages scattered across different versions and time, a few different JS libraries for web pages, etc., etc., etc.  And all of that is done several times a day shifting gears, and remotely to boot, makes it rather difficult to find much time for other things.

Hence my quest to find something "simpler" and more direct than the mess that our 3 devs have inherited.

While I understand that the bonjour/multicast DNS isn't all that "standard" - it is used in several instances, especially heterogenous networks.  Also, I believe, used by many IOT devices.

**Comment by MrMino** [comment]
I'd actually like to set up a network like that, but thanks to the pandemic I have no access to the laboratory or private hardware. I might try to set it up and check, but without specifics that won't help here.

**Comment by fangme** [comment]
Might not help - but at home (where I'm testing parts of this - and having these issues), I'm running a few Raspberry Pis.  Pi2/Pi3b/Pi3b+ and Pi Zero W.  The problem exists on all of them, and a couple of other .local devices.  A Pi Zero W is about $10USD (I got them at a store for $5USD), a "case" is about $4USD and a USB-A attachment to power it (https://shop.pimoroni.com/products/zero-stem-usb-otg-connector), appears to be around $7-8USD.  I've got 3 of them - all with that USB power, and you can also configure to connect to them over USB easily.

Bonjour is available on PC from Apple - but I have NO idea how alike it would be to a Mac running Bonjour.  But that should give you bonjour and avahi (linux) on your network fairly inexpensively.  Not like a campus environment - but will simulate parts of it.

Its just interesting that I can connect via IP 100% of the time, and using what I describe above to get the IP from a ".local" (socket.gethostbyname) name works 100% of the time.

And I've got to solve problems again.

**Comment by ma9edo** [comment]
I am running into the same problem .. the adapter IP is correct but cannot be connected!
